app/mender.go

```diff
@@ -670,3 +686,195 @@ func verifyAndSetArtifactNameInProvides(
+	log.Debugf("Validating bootstrap Artifact from %s", updatePath)
+	dt, err := m.GetDeviceType()
+	if err != nil {
+		return errors.Wrap(err, "could not get device type")
+	}
+	artifactName, artifactGroup, provides, err := validateAndParseBootstrapArtifact(dt, updatePath)
+	if err != nil {
+		return errors.Wrap(err, "invalid bootstrap Artifact")
+	}
+
+	// Initialize database
+	log.Debugf("Initializing database from bootstrap Artifact")
+	err = m.Store.WriteTransaction(func(txn store.Transaction) error {
+		return datastore.CommitArtifactData(txn, artifactName, artifactGroup, provides, nil)
+	})
+	if err != nil {
+		return errors.Wrap(err, "Could not update database")
+	}
+
+	log.Info("Bootstrap Artifact installed successfully")
+	databaseInitialized = true
+
+	return nil
+
+}
+
+func validateAndParseBootstrapArtifact(
+	dt, path string,
+) (string, string, artifact.TypeInfoProvides, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return "", "", nil, err
+	}
+	defer f.Close()
+
+	ar := areader.NewReader(f)
+
+	ar.CompatibleDevicesCallback = func(devices []string) error {
+		for _, dev := range devices {
+			if dev == dt {
+				return nil
+			}
+		}
+		return fmt.Errorf(
+			"image (device types %v) not compatible with device %v",
+			devices,
+			dt,
+		)
+	}
+
+	var scripts []string
+	ar.ScriptsReadCallback = func(r io.Reader, info os.FileInfo) error {
+		scripts = append(scripts, info.Name())
+		return nil
+	}
+
+	signed := false
+	ar.VerifySignatureCallback = func(message, sig []byte) error {
+		signed = true
+		return nil
+	}
+	err = ar.ReadArtifact()
+	if err != nil {
+		return "", "", nil, err
+	}
+
+	if ar.GetArtifactName() == "" {
+		return "", "", nil, errors.New("artifact_name cannot be emtpy")
+	}
+
+	if signed {
+		return "", "", nil, errors.New("Artifact cannot be signed")
+	}
+	if len(scripts) > 0 {
+		return "", "", nil, errors.New("Artifact cannot contain state scripts")
+	}
+
+	info := ar.GetInfo()
+	if info.Format != "mender" {
+		return "", "", nil, errors.New("wrong Artifact format")
+	}
+	if info.Version != 3 {
+		return "", "", nil, errors.New("wrong Artifact version")
+
+	}
+
+	groupDepends := ar.GetArtifactDepends()
+	if groupDepends != nil && groupDepends.ArtifactGroup != nil {
+		return "", "", nil, errors.New("artifact_depends must be emtpy")
+	}
+
+	updates := ar.GetHandlers()
+	if len(updates) != 1 {
+		return "", "", nil, errors.New("Artifact must contain exactly 1 update")
+	}
+
+	update := updates[0]
+	provides, err := update.GetUpdateProvides()
+	if err != nil {
+		return "", "", nil, errors.Wrap(err, "error reading provides")
+	}
+	if provides == nil {
+		return "", "", nil, errors.New("update must include artifact_provides")
+	}
+
+	depends, err := update.GetUpdateDepends()
+	if err != nil {
+		return "", "", nil, errors.Wrap(err, "error reading depends")
+	}
+	if depends != nil {
+		return "", "", nil, errors.New("update artifact_depends must be empty")
+	}
+
+	if len(update.GetUpdateAllFiles()) != 0 {
```

Comment:
Q: Not against being careful, but isn't this something the library should already guard against?

Reply:
How exactly? If you create a bootstrap Artifact yes, the library guard against it, but for any other Artifact it wont. I could craft an Artifact with type `nil` but still have actual payload (the current code will anyway ignore the payload, but I think is good to be proactive here).

Reply:
I understand :)